gh-133885: Use locks instead of critical sections for _zstd

emmatyping opened the pull request:
Move from using critical sections to locks for the (de)compression methods. Since the methods allow other threads to run, we should use a lock rather than a critical section.

I'm not totally sure about the removal of the critical sections in the `_get_(C|D)Dict` function, but it should be safe to yield to other threads in an initializer as the values cannot be modified (ZstdDicts are immutable).



<!-- gh-issue-number: gh-133885 -->
* Issue: gh-133885
<!-- /gh-issue-number -->

emmatyping committed:
Use locks instead of critical sections

Also remove critical sections around code only run in initializer.

emmatyping committed:
Remove one more critical section

ZeroIntensity commented on Modules/_zstd/compressor.c:
```diff
@@ -403,6 +399,10 @@ ZstdCompressor_dealloc(PyObject *ob)
 
+    if (PyMutex_IsLocked(&self->lock)) {
+        PyMutex_Unlock(&self->lock);
+    }
```
This pattern will cause a fatal error if people are using the locks correctly.

colesbury replied:
Yeah, we should avoid this pattern. You can do something like:

```
assert(!PyMutex_IsLocked(&self->lock));
```

at the beginning of `ZstdCompressor_dealloc`, but we should structure the locking such that it's impossible for the compressor to be deallocated while the lock is held.

emmatyping replied:
Yeah, that was what I was thinking, getting in that state should be impossible, but good to check in case it does happen somehow.

ZeroIntensity replied:
If we reach the destructor without releasing the lock, that's a good sign that something went horribly wrong, especially if we make sure the lock can't be held when calling into Python. I'd honestly rather a deadlock than letting it silently pass.

ZeroIntensity reviewed (commented):
This seems like a downgrade. Though rare, functions like `PyMem_Malloc` can call arbitrary Python code, so we lose our protection against deadlocks. What's the rationale here?

colesbury commented on Lib/test/test_zstd.py:
```diff
@@ -2430,7 +2430,6 @@ def test_buffer_protocol(self):
-@unittest.skip("it fails for now, see gh-133885")
 class FreeThreadingMethodTests(unittest.TestCase):
 
     @unittest.skipUnless(Py_GIL_DISABLED, 'this test can only possibly fail with GIL disabled')
```
We should remove this `@unittest.skipUnless(Py_GIL_DISABLED, ...);`. It's very possible for the test to fail with the GIL enabled.

emmatyping replied:
Removed this, it's a good point!

colesbury commented on Modules/_zstd/compressor.c:
```diff
@@ -38,6 +39,9 @@ typedef struct {
     /* Compression level */
     int compression_level;
+
+    /* Lock to protect the compression context */
+    PyMutex lock;
```
I think this is a right idea. We need a lock for both the GIL-enabled and FT builds, because the GIL doesn't protect things within a `Py_BEGIN_ALLOW_THREADS`, which is used before many ZSTD API calls.

colesbury commented on Modules/_zstd/compressor.c:
```diff
@@ -153,8 +157,6 @@ _get_CDict(ZstdDict *self, int compressionLevel)
     PyObject *capsule;
     ZSTD_CDict *cdict;
 
-    // TODO(emmatyping): refactor critical section code into a lock_held function
```
I think `_get_CDict` needs some more work. It doesn't look thread-safe to me (even with the GIL) because the `Py_BEGIN_ALLOW_THREADS` below means multiple threads can try to create the capsule for the same level.

I can think of two high-level strategies:

1) Use a lock on `ZstdDict` to ensure that only one thread tries to initialize the `cdict` at a time.

2) Use `PyDict_SetDefaultRef` so that only one thread succeeds in setting the capsule on the dict.

Is `self->dict_content` mutable? If so, then the lock may be more appropriate.

emmatyping replied:
Yeah that's a great point! I think `PyDict_SetDefaultRef` sounds more reasonable, the content is immutable so it should be safe to use `PyDict_SetDefaultRef`.

emmatyping replied:
I've switched to using PyDict_SetDefaultRef and PyDict_GetItemRef. I also added a test that should check sharing a dictionary across threads with several levels.

emmatyping replied:
Okay I actually used locking in the end, because we also want to make sure the member access of `ZstdDict->d_dict` is thread safe.

emmatyping committed:
Make capsule management threadsafe

Using PyDict_SetDefaultRef allows us to not have contention on the
capsules dictionary shared by ZstdDict instances.

emmatyping committed:
Add a test for sharing a dictionary across threads

emmatyping committed:
Move lock check in dealloc to an assert

emmatyping committed:
Run threading checks on GIL builds too

emmatyping committed:
Remove extra blank line

emmatyping committed:
Remove extraneous whitespace

ZeroIntensity commented on Lib/test/test_zstd.py:
```diff
@@ -2506,6 +2503,27 @@ def run_method(method, input_data, output_data):
+            # sync threads to increase chance of contention on
+            # capsule storing dictionary levels
+            b.wait()
+            ZstdCompressor(level=level, zstd_dict=TRAINED_DICT.as_digested_dict)
```
It'd be nice if we tested for sharing a compressor between threads too.

emmatyping replied:
We do that above in `test_compress_locking`, unless you had something else in mind?

ZeroIntensity replied:
Oh, didn't see that.

ZeroIntensity commented on Modules/_zstd/decompressor.c:
```diff
@@ -61,7 +65,6 @@ _get_DDict(ZstdDict *self)
-    Py_BEGIN_CRITICAL_SECTION(self);
     if (self->d_dict == NULL) {
         /* Create ZSTD_DDict instance from dictionary content */
         char *dict_buffer = PyBytes_AS_STRING(self->dict_content);
```
Similar to the other case, we need to lock here to prevent races on initializing `self->d_dict` with multiple threads.

emmatyping replied:
Yeah, I should probably put locks around `ZstdDict` to protect here and in `_get_CDict`.

emmatyping replied:
Okay, hopefully the new changes look good!

colesbury replied:
GitHub won't let me comment on the actual line, but I think it's good practice to avoid modifying Python object fields within `Py_BEGIN_ALLOW_THREADS`. Instead of:

```c
        Py_BEGIN_ALLOW_THREADS
        self->d_dict = ZSTD_createDDict(dict_buffer,
                                        dict_len);
        Py_END_ALLOW_THREADS
```

use something like:

```c
        PyObject *d_dict;
        Py_BEGIN_ALLOW_THREADS
        d_dict = ZSTD_createDDict(dict_buffer, dict_len);
        Py_END_ALLOW_THREADS
        self->d_dict = d_dict;
```

I think the current code is probably safe, but it's easy to have subtle data races with this sort of pattern. For example, I think it would not be safe if you were instead assigning to `self->c_dicts` because that field can be read by a `tp_traverse` concurrently with a `Py_BEGIN_ALLOW_THREADS` block (in both the GIL-enabled and free-threaded builds).

emmatyping replied:
Excellent point, I've updated this to avoid that.

ZeroIntensity commented on Modules/_zstd/decompressor.c:
```diff
@@ -639,7 +635,10 @@ _zstd_ZstdDecompressor_unused_data_get_impl(ZstdDecompressor *self)
 {
```
The `@critical_section` in the AC definition isn't needed anymore. We probably need to do that in a couple other places too.

colesbury replied:
`_zstd.ZstdDecompressor.unused_data` still has the `@critical_section`.

emmatyping replied:
That should now be removed, thanks for catching it!

ZeroIntensity reviewed (commented):
Ok, I understand this a little better now, sorry!

I'm a bit iffy about the lock-ordering here. I *think* the lock is only held under non-reentrant calls? `PyErr_SetString` and friends seem to be safe, but it looks like they might be able to re-enter if someone did weird things to the exception state. I wouldn't worry about it, really, but it's worth bringing up.

emmatyping committed:
Remove unused includes

emmatyping committed:
Revert "Remove unused includes"

This reverts commit 005d1a06ba8cd1fcef01f4fbcc57b010657f0f7e.

emmatyping committed:
Put locks around ZstdDict usage

This also reverts the setdefault usage which we don't need anymore and refactors the _zstd_load_(c,d)_dict functions to not use goto/properly lock around dictionary usage.

emmatyping committed:
Add tests for sharing all types of ZstdDict

emmatyping committed:
Remove @critical_section

emmatyping committed:
Add missing include

colesbury commented on Modules/_zstd/compressor.c:
```diff
@@ -195,7 +196,7 @@ _get_CDict(ZstdDict *self, int compressionLevel)
         }
 
-        /* Add PyCapsule object to self->c_dicts */
+        /* Add PyCapsule object to self->c_dicts if not already inserted */
```
I think you should use `PyDict_SetDefaultRef` instead of `PyDict_SetItem` below

emmatyping replied:
Done! I didn't realize the semantics around SetItem were quite so weird.

colesbury replied:
I'm sorry - I was confused and forgot this function is now called with the `ZstdDict` lock held. `PyDict_SetItem` was fine here.

emmatyping replied:
Ah, no worries! I don't think using PyDict_SetDefaultRef is wrong per se. But I can always undo that change.

colesbury replied:
Yeah, I don't think it's wrong, just maybe more complicated than necessary. Up to you.

emmatyping replied:
I think I will revert to PyDict_SetItem. Do you have thoughts on using PyDict_GetItemRef here: https://github.com/python/cpython/pull/134289#discussion_r2102566026 ?

I would think similarly it's not really needed.

colesbury replied:
`PyDict_GetItemWithError` is safe here because we only add new entries to `c_dicts`, and never remove or update them, but there is now generally a preference for APIs that return new references over borrowed references.

I think it's fine to switch to `PyDict_GetItemRef` -- it avoids having to worry about why borrowed references are safe in this context.

colesbury commented on Modules/_zstd/decompressor.c:
```diff
@@ -83,7 +86,6 @@ _get_DDict(ZstdDict *self)
 
     /* Don't lose any exception */
     ret = self->d_dict;
-    Py_END_CRITICAL_SECTION();
```
This can be simplified now to just `return self->d_dict;`

colesbury commented on Modules/_zstd/decompressor.c:
```diff
@@ -61,7 +65,6 @@ _get_DDict(ZstdDict *self)
         return self->d_dict;
```
Up to you, but I'd consider adding something like a `assert(PyMutex_IsLocked(&self->lock));` at the beginning of `_get_DDict`

colesbury reviewed (commented):
Looks pretty good to me. A few small comments below

emmatyping committed:
Updates based on Sam's review

- Assert locks are held in `_get_(C/D)Dict`
- Update self->d_dict outside `Py_BEGIN_ALLOW_THREADS`
- removed critical section AC I missed

ZeroIntensity commented on Modules/_zstd/compressor.c:
```diff
@@ -149,12 +153,11 @@ capsule_free_cdict(PyObject *capsule)
 ZSTD_CDict *
 _get_CDict(ZstdDict *self, int compressionLevel)
 {
+    assert(PyMutex_IsLocked(&self->lock));
```
Ouch, we need to be really careful about this lock. Calls like `PyDict_GetItemWithError` or `Py_DECREF` can definitely call Python code, so it's cloudy with a chance of deadlocks. It *might* be fine--is `c_dicts` ever exposed to Python?

emmatyping replied:
cdicts holds compression dictionary data and is only mutated through internal functions, so no, it is only ever a normal dictionary.

emmatyping replied:
more precisely it stores zstd compression dictionaries (bytes) in a Python dictionary keyed by compression level

emmatyping replied:
I agree the potential for deadlocks are concerning, perhaps it would be good to note this invariant in a comment?

kumaraditya303 replied:
This may not be an issue here but `PyDict_GetItemRef` should be preferred over `PyDict_GetItemWithError` as latter returns borrowed reference which isn't safe.

ZeroIntensity replied:
Ok, yeah, let's use `PyDict_GetItemRef` while we're here. As long as `c_dicts` definitely isn't re-entrant, there shouldn't be any deadlocks.

ZeroIntensity reviewed (commented):
Yay, buildbots passed! LGTM too, apart from one (speculative) concern about re-entrancy.

kumaraditya303 commented on Modules/_zstd/compressor.c:
```diff
@@ -195,12 +197,15 @@ _get_CDict(ZstdDict *self, int compressionLevel)
+        PyObject *capsule_dict;
+        int ret = PyDict_SetDefaultRef(self->c_dicts, level, capsule,
+                                       &capsule_dict);
+        Py_XDECREF(capsule_dict);
```
```suggestion
        int ret = PyDict_SetDefaultRef(self->c_dicts, level, capsule, NULL);
```
You can pass NULL since `capsule_dict` is not needed

kumaraditya303 commented on Modules/_zstd/compressor.c:
```diff
@@ -412,8 +422,8 @@ ZstdCompressor_dealloc(PyObject *ob)
-compress_impl(ZstdCompressor *self, Py_buffer *data,
-              ZSTD_EndDirective end_directive)
+compress_lock_held(ZstdCompressor *self, Py_buffer *data,
+                   ZSTD_EndDirective end_directive)
```
Add `assert(PyMutex_IsLocked(&self->lock));` to lock held functions to ensure the lock is held.

kumaraditya303 commented on Modules/_zstd/compressor.c:
```diff
@@ -495,7 +505,7 @@ mt_continue_should_break(ZSTD_inBuffer *in, ZSTD_outBuffer *out)
 static PyObject *
-compress_mt_continue_impl(ZstdCompressor *self, Py_buffer *data)
+compress_mt_continue_lock_held(ZstdCompressor *self, Py_buffer *data)
 {
```
Same

emmatyping committed:
Pass NULL to PyDict_SetDefaultRef

Co-authored-by: Kumar Aditya <kumaraditya@python.org>

emmatyping committed:
Put locks around relevant functions and use PyDict_GetItemRef

ZeroIntensity commented on Modules/_zstd/compressor.c:
```diff
@@ -163,12 +167,12 @@ _get_CDict(ZstdDict *self, int compressionLevel)
-    capsule = PyDict_GetItemWithError(self->c_dicts, level);
+    ret = PyDict_GetItemRef(self->c_dicts, level, &capsule);
+    if (ret < 0) {
+        Py_XDECREF(capsule);
```
Won't `capsule` always be `NULL` here?

emmatyping replied:
Yes, good point! I'll remove this.

emmatyping committed:
Remove unused decref

ZeroIntensity reviewed (approved):
LGTM assuming buildbots pass
